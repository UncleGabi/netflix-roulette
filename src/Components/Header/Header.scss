.header {
  position: relative;
  height: 500px;
  margin-bottom: 10px;
  background: linear-gradient(rgba(0, 0, 0, 0.7), rgba(0, 0, 0, 0.7)),
    url("../../Assets/images/header-bg.jpg");
  background-size: cover;
  background-position: center;
  background-repeat: no-repeat;

  &__title {
    display: flex;
    justify-content: space-between;
    padding: 15px 40px;
    color: #f65261;
    z-index: 2;

    &-logo {
      font-size: 20px;

      span {
        font-weight: bold;
      }
    }

    button {
      width: 180px;
      height: 45px;
      background: #606060ad;
      color: #f65261;
      font-size: 20px;
      text-transform: uppercase;
      border: 1px solid #606060ad;
      cursor: pointer;
    }
  }

  &__search-container {
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    width: fit-content;
    color: white;

    h2 {
      font-size: 30px;
      font-weight: 300;
      margin-bottom: 30px;
      text-transform: uppercase;
    }

    input {
      width: 650px;
      height: 55px;
      margin-right: 10px;
      padding-left: 20px;
      background: rgba($color: #323232cc, $alpha: 0.8);
      color: #fff;
      font-size: 20px;
      border: none;
      outline: none;
    }

    button {
      width: 235px;
      height: 55px;
      background-color: #f65261;
      color: #fff;
      font-size: 20px;
      border: none;
      text-transform: uppercase;
      cursor: pointer;
    }
  }
}
