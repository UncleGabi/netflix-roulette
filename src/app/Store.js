import { configureStore } from "@reduxjs/toolkit";
import MoviesReducer from "../features/MoviesSlice";

export const store = configureStore({
  reducer: {
    movies: MoviesReducer,
  },
});
