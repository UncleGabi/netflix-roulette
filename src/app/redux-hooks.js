import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";

export const useAppDispatch = useDispatch;
export const useAppSelector = useSelector;
export const useAppNavigate = useNavigate;
