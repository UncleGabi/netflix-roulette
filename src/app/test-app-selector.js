const state = {
  movies: [],
  filteredMovies: [],
  selectedMovie: {},
  editedMovie: {},
  searchMovieTitle: "",
};

export const testUseAppSelector = (f) => f(state);
