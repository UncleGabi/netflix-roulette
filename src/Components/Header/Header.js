import React, { useEffect, useState } from "react";
import { useSearchParams } from "react-router-dom";
import { v4 as uuidv4 } from "uuid";
import {
  addMovie,
  selectAllMovies,
  setEditedMoive,
  setSearchMovieTitle,
} from "../../features/MoviesSlice";
// import { useDispatch, useSelector } from "react-redux";
import { useAppDispatch, useAppSelector } from "../../app/redux-hooks";

import Modal from "../Common/Modal/Modal";
import Form from "../Common/Modal/Form/Form";

import "./Header.scss";

const Header = () => {
  const [openModal, setOpenModal] = useState(false);
  const [searchParams, setSearchParams] = useSearchParams();
  const [title, setTitle] = useState(searchParams.get("title") || "");
  const dispatch = useAppDispatch();
  const { editedMovie, searchMovieTitle } = useAppSelector(selectAllMovies);

  const handleChange = (e) => {
    setTitle(e.target.value);
    dispatch(setSearchMovieTitle(e.target.value));
  };

  const handleSearch = () => {
    const currentParams = Object.fromEntries(searchParams.entries());
    setSearchParams({ ...currentParams, title: searchMovieTitle });
  };

  return (
    <header className="header">
      <div className="header__title">
        <div className="header__title-logo">
          <span>netflix</span>roulette
        </div>
        <button
          onClick={() => {
            setOpenModal(true);
            dispatch(
              setEditedMoive({
                title: "",
                vote_average: 0,
                genres: [],
                release_date: new Date().toLocaleDateString(),
                runtime: 0,
                overview: "",
              })
            );
          }}
        >
          + Add movie
        </button>
      </div>
      <div className="header__search-container">
        <h2>Find your movie</h2>
        <div>
          <input
            type="text"
            placeholder="What do you want to watch?"
            value={title}
            onChange={handleChange}
            onKeyDown={(e) => {
              if (e.key === "Enter") {
                if (title || searchParams.get("title")) {
                  handleSearch();
                }
              }

              if (e.key === "Escape") {
                setTitle("");
              }
            }}
            autoFocus
          />
          <button id="search" onClick={handleSearch}>
            Search
          </button>
        </div>
      </div>
      {openModal && (
        <Modal
          title="Add movie"
          width={900}
          height={750}
          primaryButtonFn={() => {
            dispatch(
              addMovie({
                id: uuidv4(),
                ...editedMovie,
                release_date:
                  editedMovie.release_date || new Date().toLocaleDateString(),
              })
            );
            setOpenModal(undefined);
          }}
          secondaryButtonFn={() => setOpenModal(undefined)}
          setOpenModal={setOpenModal}
        >
          <Form />
        </Modal>
      )}
    </header>
  );
};

export default Header;
