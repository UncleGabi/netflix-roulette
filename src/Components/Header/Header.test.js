/* eslint-disable testing-library/no-node-access */
/* eslint-disable testing-library/no-container */
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { useAppDispatch, useAppSelector } from "../../app/redux-hooks";
import renderer from "react-test-renderer";

import Header from "./Header";
import { BrowserRouter as Router } from "react-router-dom";
import { testUseAppSelector } from "../../app/test-app-selector";

jest.mock("../../app/redux-hooks");

describe("Header component", () => {
  const dispatch = jest.fn;

  beforeEach(() => {
    useAppSelector.mockImplementation(testUseAppSelector);
    useAppDispatch.mockImplementation(dispatch);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  const renderHeader = () => (
    <Router>
      <Header />
    </Router>
  );

  it("should render Header component correctly", () => {
    const tree = renderer
      .create(
        <Router>
          <Header />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should change input value when user types in it", () => {
    const { container } = render(
      <Router>
        <Header />
      </Router>
    );
    const titleInput = container.querySelector("input");

    userEvent.type(titleInput, "movie");

    expect(titleInput).toHaveValue("movie");
    expect(useAppDispatch).toHaveBeenCalled();
  });
});
