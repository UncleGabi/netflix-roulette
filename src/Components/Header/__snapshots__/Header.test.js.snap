// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component should render Header component correctly 1`] = `
<header
  className="header"
>
  <div
    className="header__title"
  >
    <div
      className="header__title-logo"
    >
      <span>
        netflix
      </span>
      roulette
    </div>
    <button
      onClick={[Function]}
    >
      + Add movie
    </button>
  </div>
  <div
    className="header__search-container"
  >
    <h2>
      Find your movie
    </h2>
    <div>
      <input
        autoFocus={true}
        onChange={[Function]}
        onKeyDown={[Function]}
        placeholder="What do you want to watch?"
        type="text"
        value=""
      />
      <button
        id="search"
        onClick={[Function]}
      >
        Search
      </button>
    </div>
  </div>
</header>
`;
